src/NHibernate.Test/QueryTest/QueryBatchFixture.cs

```diff
@@ -0,0 +1,1003 @@
+		{
+			CreateItems();
+
+			using (var session = OpenSession())
+			{
+				var transformer = new ResultTransformerStub();
+				var query = session.CreateSQLQuery("select * from ITEM")
+				                   .AddEntity(typeof(Item))
+				                   .SetResultTransformer(transformer);
+				session.CreateQueryBatch()
+				       .Add<object[]>(query)
+				       .GetResult<object[]>(0);
+
+				Assert.That(transformer.WasTransformTupleCalled, Is.True, "Transform Tuple was not called");
+				Assert.That(transformer.WasTransformListCalled, Is.True, "Transform List was not called");
+			}
+		}
+
+		[Test]
+		public void CannotAddMixedQueryWithKeyThatAlreadyExists()
+		{
+			using (var session = OpenSession())
+			{
+				var queries = session.CreateQueryBatch();
+
+				var criteria = DetachedCriteria.For(typeof(Item))
+				                               .Add(Restrictions.Lt("id", 50));
+
+				var query = session.Query<Item>();
+
+				queries.Add<Item>("commonKey", criteria);
+
+				Assert.That(
+					() => queries.Add("commonKey", query),
+					Throws.ArgumentException);
+			}
+		}
+
+		[Test]
+		public void CannotRetrieveResultWithUnknownKey()
+		{
+			CreateItems();
+
+			using (var session = OpenSession())
+			{
+				var multiCriteria = session.CreateQueryBatch();
+
+				var firstCriteria = session.CreateCriteria(typeof(Item))
+				                           .Add(Restrictions.Lt("id", 50));
+
+				multiCriteria.Add<Item>("firstCriteria", firstCriteria);
+
+				Assert.That(
+					() => multiCriteria.GetResult<Item>("unknownKey"),
+					Throws.InstanceOf<KeyNotFoundException>());
+			}
+		}
+
+		private void DoMultiQueryAndAssert()
+		{
+			using (var s = OpenSession())
+			{
+				var multiQuery = s.CreateQueryBatch()
+				                  .Add<Item>(
+					                  s.CreateSQLQuery("select * from ITEM where Id > ?")
+					                   .AddEntity(typeof(Item))
+					                   .SetInt32(0, 50)
+					                   .SetFirstResult(10)
+					                   .SetCacheable(true))
+				                  .Add<long>(
+					                  s.CreateQuery("select count(*) from Item i where i.Id > ?")
+					                   .SetInt32(0, 50)
+					                   .SetCacheable(true));
+				var items = multiQuery.GetResult<Item>(0);
+				Assert.That(items.Count, Is.EqualTo(89));
+				var count = multiQuery.GetResult<long>(1).Single();
+				Assert.That(count, Is.EqualTo(99L));
+			}
+		}
+
+		#endregion
+
+		private void CreateItems()
+		{
+			using (var s = OpenSession())
+			using (var t = s.BeginTransaction())
+			{
+				for (var i = 0; i < 150; i++)
+				{
+					var item = new Item { Id = i };
+					s.Save(item);
+				}
+
+				t.Commit();
+			}
+		}
+
+		/// <summary>
+		/// Get the inner Hashtable from the IQueryCache.Cache
+		/// </summary>
+		/// <returns></returns>
+		public static Hashtable GetHashTableUsedAsQueryCache(ISessionFactoryImplementor factory)
```

Comment:
Could you please clarify the reasons for this hackish method? It seems all usages are equivalent to `factory.QueryCache.Clear()`

Reply:
No I cannot. That is just the way it was done for multi-criteria/query.

Reply:
In fact reading the code, I can: it is not used only for clearing, but also for checking directly that it gets filled with items.

Reply:
Isn't it something that can be checked with "Sfi.Statistics.QueryCachePutCount"?

Reply:
It could, assuming there are no bugs in stats. Here with this test it is really certain things are in the cache.

Reply:
But that's something that you are already using for other tests -  `CanGetMultiCriteriaFromSecondLevelCache`  for example.
So the whole usage seems to be analog of clearing statistics and checking put counts. IMHO we should get rid of this method for good.

Reply:
Yes I know it is done for other tests. About tests I do not think consistency in the way things are tested is good. Instead, testing with different ways help catching more cases.

Reply:
I see your point. But for me it looks like introducing dirty hack for no particular reasons (to me checking count of items in internal cache implementation doesn't add any value to existing tests - you can't say for sure that those items are stored properly and can be retrieved by the same query reexecution).